framework/materials/multi_group_xs/openmc_import.cc

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "framework/materials/multi_group_xs/multi_group_xs.h"
+#include "framework/logging/log.h"
+#include "framework/utils/hdf_utils.h"
+#include <numeric>
+#include <algorithm>
+
+namespace opensn
+{
+
+void
+MultiGroupXS::Initialize(const std::string& file_name, double temperature)
+{
+  // Lambda for checking for all non-negative values.
+  auto IsNonNegative = [](const std::vector<double>& vec)
+  { return not std::any_of(vec.begin(), vec.end(), [](double x) { return x < 0.0; }); };
+
+  // Lambda for checking for all strictly positive values.
+  auto IsPositive = [](const std::vector<double>& vec)
+  { return not std::any_of(vec.begin(), vec.end(), [](double x) { return x <= 0.0; }); };
+
+  // Lambda for checking for any non-zero values.
+  auto HasNonZero = [](const std::vector<double>& vec)
+  { return std::any_of(vec.begin(), vec.end(), [](double x) { return x > 0.0; }); };
+
+  Reset();
+
+  // Disable internal HDF error reporting
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+
+  // Open file
+  try
+  {
+    H5::H5File::isHdf5(file_name.c_str());
+  }
+  catch (H5::FileIException(&err))
+  {
+    std::string err_msg = file_name + " is not an HDF5 file.\n";
+    throw std::logic_error(err_msg);
+  }
+  H5::H5File file(file_name.c_str(), H5F_ACC_RDONLY);
+
+  // Check file type
+  std::string filetype;
+  H5ReadAttribute<std::string>(file, "filetype", filetype);
+  if (filetype != "mgxs")
+    throw std::runtime_error(file_name + " is not a valid OpenMC library file");
+
+  log.Log() << "Reading OpenMC cross section file \"" << file_name << "\"\n";
+
+  // Number of groups
+  if (not H5ReadAttribute<size_t>(file, "energy_groups", num_groups_))
+    throw std::runtime_error("Failure reading \"energy_groups\" from " + file_name);
+
+  // Group structure
+  e_bounds_ = H5ReadDataset1D<double>(file, "/group structure");
+  std::reverse(e_bounds_.begin(), e_bounds_.end());
+
+  // Temperature
+  log.Log0() << file_name + " cross-section data evaluated at " << temperature << "K\n";
+  temperature_ = temperature;
+
+  // Base path
+  std::stringstream ss;
+  ss << std::fixed << std::setprecision(0) << temperature_ << "K";
+  std::string path = "/set1/" + ss.str() + "/";
```

Comment:
A library can have multiple sets of cross sections. Right now, I only support one per file. We can expand on this later as needed.

Reply:
I see. I assume this is a file path here? If so, I think we should avoid hardcoding into a path. Could this be abstracted in any way?

Reply:
I'm not sure what you mean by file path. It's the location of the dataset internal to the xs library. It's not something I would really expect the user to know anything about.  Any xs file should have at least one set and one temperature. We only support single material/single temperature libraries for the time being.  There should always be a `set1` in the library. I can make a change to generate a warning if we find multiple sets.

Reply:
Without knowing much HDF5 stuff the comment "// Base path" read as a file path to me, not an internal HDF5 path... I see now. I'll revise the question with this in mind... If I want to generate XS from OpenMC, will the library have the `set1` path by default, or is this something that is set by the user when generating the XS? If `set<n>` is immutable on the OpenMC side, then this is good. If it is a user-settable label on the OpenMC side, our reader should reflect that.

Reply:
:smile: Yeah, I can see how `base path` could be misleading. Since the internals of an HDF file remind me of a file system, I guess that's where my brain goes. The `path` including `set<n>` is not something that the user has any control over. The only part of the `path` that is influenced by the user is the temperature.

Reply:
Okay, perfect! Thanks!

Reply:
Hey @zhardy-lanl. Thanks to your prodding, I looked into this more closely. It is possible for the user to specify the dataset names in the HDF file. If not specified at library creation, it defaults to `set<n>`. I've pushed changes to take an optional argument that specifies the dataset name. It defaults to `set1`, but we can now have multiple xs datasets in a single file and the user and assign them as appropriate. 👍

Reply:
I've made some minor fixes to support custom set names and added an associated regression test.